coeus-impl/src/main/java/org/kuali/coeus/propdev/impl/attachment/ProposalDevelopmentAttachmentHelper.java

```diff
@@ -38,6 +44,163 @@ public void reset() {
+    	// Delete Narratives which are not on the form - from document.
+    	if(documentBios != null && documentBios.size() > 0) {
+    		List<ProposalPersonBiography> deletedBios = new ArrayList<ProposalPersonBiography>();
+    		for(ProposalPersonBiography personBio : documentBios ) {
+    			if(findPersonBiosByObjectId(formBios, personBio).size() == 0 ) {
+    				deletedBios.add(personBio);
+            	}
+    		}
+	   		if(deletedBios.size() > 0) {
+	   			documentBios.removeAll(deletedBios);
+	   		}
+    	}
+    }
+    
+    private void addNewUpdatePersonBiosOnDocument(List<ProposalPersonBiography> formBios, List<ProposalPersonBiography> documentBios) {
+	   	
+    	for(ProposalPersonBiography personBio :  formBios) {
+	   		if(personBio.getObjectId() == null) {
+	   			documentBios.add(personBio);
+	   			continue;
+	   		}
+	       	if(personBio.isUpdated()) {
+	       		ProposalPersonBiography updatedBio = findUpdatedPersonBio(documentBios,personBio);
+	       		documentBios.remove(documentBios.indexOf(updatedBio));
+	       		documentBios.add(updatedBio);
+	       	}
+	   	}
+    }
+    
+    private void addNewUpdateNarrativesOnDocument(List<Narrative> formNarratives, List<Narrative> documentNarratives) {
+	   	
+    	for(Narrative narrative :  formNarratives) {
+	   		if(narrative.getObjectId() == null) {
+	   			documentNarratives.add(narrative);
+	   			continue;
+	   		}
+	       	if(narrative.isUpdated()) {
+	       		Narrative updatedNarrative = findUpdatedNarrative(documentNarratives,narrative);
+	       		documentNarratives.remove(documentNarratives.indexOf(updatedNarrative));
+	       		documentNarratives.add(narrative);
+	       	}
+	   	}
+    }
+    	
+    private void removeDeletedNarrativesFromDocument( List<Narrative> formNarratives, List<Narrative> documentNarratives) {
+    	// Delete Narratives which are not on the form - from document.
+    	if(documentNarratives != null && documentNarratives.size() > 0) {
+    		List<Narrative> deletedNarratives = new ArrayList<Narrative>();
+    		for(Narrative narrative : documentNarratives ) {
+    			if(findNarrativesByType(formNarratives, narrative.getNarrativeTypeCode()).size() == 0 ) {
+            		deletedNarratives.add(narrative);
+            	}
+    		}
+	   		if(deletedNarratives.size() > 0) {
+	   			documentNarratives.removeAll(deletedNarratives);
+	   		}
+    	}
+    }
+    
+    protected ProposalPersonBiography findUpdatedPersonBio(List<ProposalPersonBiography> origBios, final ProposalPersonBiography newBio) {
+    	
+    	ProposalPersonBiography updatedBio = CollectionUtils.find(origBios, new Predicate<ProposalPersonBiography>() {
+
+			@Override
+			public boolean evaluate(ProposalPersonBiography input) {
+				if(input.getObjectId().equals(newBio.getObjectId())) {
+					return true;
+				}	
+				else {
+					return false;
+				}
```

Comment:
And objectId here. Prefer use of the primary key instead. And actually looking at this again, aren't these find methods essentially duplicates? In one case you are returning a unique result and in others a list, but in both cases you are basing the search on the same criteria.